Configurable password hashing algorithm/cost

jkakavas opened the pull request:
for the stored passwords of users for the realms that this applies
(native, reserved). Replaces predefined choice of bcrypt with
cost factor 10.
This also introduces PBKDF2 with configurable cost
(number of iterations) as an algorithm option for password hashing
both for storing passwords and for the user cache.
Doesn't support "on the fly" change of hashing algorithm selection
as pre-existing users won't be able to authenticate.

Documentation additions will be handled in a separate PR.

elasticmachine commented:
Pinging @elastic/es-security

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackSettings.java:
```diff
@@ -106,6 +107,19 @@
+    /*
+     * Do not allow insecure hashing algorithms to be used for password hashing
+     */
+    public static final Setting<String> PASSWORD_HASHING_ALGORITHM = new Setting<>(
```
This part has not been discussed, but it is a good idea for a future enhancement. I'd rather wait on opening this up though; once we do that then it has to be supported by us and everything we add creates overhead. We also do not know if there is demand for this and I'd guess that the majority of users would not have a preference as long as we use something that is secure.

jkakavas replied:
This PR was in the context of supporting a FIPS-140 compliant solution, hence only PBKDF2 was added. I also haven't seen any argon2 or scrypt JAVA implementations that have been used / tested sufficiently  (there are bindings for the Argon2 C implementation though.. ) 

I'm definitely not against allowing for extensions with other algorithm implementations, but not in the context of this PR.

albertzaharovits commented:
Rant inbound:

I don't like the flow, specifically I don't like that whenever there's a password there must be a hash name tag close by. I understand why you did it, to make it testable, but I don't think this is how it should work in the end.
I also don't like that the hashing algorithm and the cost are separate settings. The range of values of the cost for one algorithm is not related to another one.
Also, `CACHE_HASH_ALGO_SETTING` should default to `PASSWORD_HASHING_ALGORITHM`. I think this realm setting is useless (or let's say, power user focused) but given the FIPS requirements I see it as dangerous. I know, requirements are different for in-memory compared to on disk hashes (they should be faster, that's why the cache is there), but should we really use `sha256 `for most password hash verifies and only the stored hashes will be `PBKDF2` ?

Instead, this is how I see things:
The `UserAndPassword` entity can also harbor the hash algo tag. This would be picked up, by looking at the hash prefix. When you need to verify the password you use the hash tag of the retrieved hash, instead of using the setting value. The setting value is used when writing the hash entry (i.e. put user or change password).

tvernum commented:
> I also don't like that the hashing algorithm and the cost are separate settings. The range of values of the cost for one algorithm is not related to another one.

+1

That decision leads to all sorts of weirdness in the code, like allowing `bcrypt10` as an algorithm, but ignoring the cost factor if one is explicitly set. Except for the cache, where you can only set it in the algorithm, because there's no cost factor.
But there's no support for using PBKDF2 with a custom cost factor for the cache algo.

albertzaharovits commented:
My previous comment was stern and unhelpful, apologies. Let me have another try at it:

GH does not allow for comments on lines that have not been changed, so here is the gist of my previous comment:
```
     void verifyPassword(String username, final SecureString password, ActionListener<AuthenticationResult> listener) {
         getUserAndPassword(username, ActionListener.wrap((userAndPassword) -> {
             if (userAndPassword == null || userAndPassword.passwordHash() == null) {
                 listener.onResponse(AuthenticationResult.notHandled());
             // HERE: can we have `userAndPassword.verify(password)`?
             } else if (hasher.verify(password, userAndPassword.passwordHash())) {
                 listener.onResponse(AuthenticationResult.success(userAndPassword.user()));
             } else {
                 listener.onResponse(AuthenticationResult.unsuccessful("Password authentication failed for " + username, null));
             }
         }, listener::onFailure));
}
```

jkakavas commented:
> That decision leads to all sorts of weirdness in the code like allowing bcrypt10 as an algorithm, but ignoring the cost factor if one is explicitly set.

Since I went down that way I had to make a decision on what to do when the cost is set both explicitly and implicitly 

```
PASSWORD_HASHING_ALGORITHM bcrypt10
PASSWORD_HASHING_COST 11
```
would give you bcrypt with cost factor 11. 

> But there's no support for using PBKDF2 with a custom cost factor for the cache algo.

The threat model doesn't usually include offline brute force attacks against the hashed passwords in the cache, thus I assumed the default cost would be sufficient for this. 

As said I do see the point at hand, thoughts about
>I see your point. The idea is that I didn't want to handle all possible cost options in the algorithm settings (i.e, bcrypt4 to bcrypt16 is ok but BKDF2_1000 to PBKDF2_100000 not that much). I have no objections removing the cost setting but maybe limit the support to ~10 costs for Bcrypt and ~5 for PBKDF2 , 

WDYT ?

albertzaharovits commented:
> The idea is that I didn't want to handle all possible cost options in the algorithm settings (i.e, bcrypt4 to bcrypt16 is ok but BKDF2_1000 to PBKDF2_100000 not that much). I have no objections removing the cost setting but maybe limit the support to ~10 costs for Bcrypt and ~5 for PBKDF2 . 

I see... validating hash algorithms is equivalent to validating cost factors. Besides pegging them, i.e. having a list of predefined ones, I think the only way is to just try and time the hash function? If it takes longer than 100 msec that we don't support the alg+cost combination? This would be a setting validation function. Sounds wasteful and insubstantial, as the timing at the validation time, for a single value might not reflect the real life timing when multiple requests are inbound.
I am leaning towards having a set of common alg+cost pairs and white list only these.

jkakavas commented:
Thanks @albertzaharovits , much more clear now. So your suggestion is to change
```
 if (userAndPassword == null || userAndPassword.passwordHash() == null) {
     listener.onResponse(AuthenticationResult.notHandled());
 } else if (hasher.verify(password, userAndPassword.passwordHash())) {
     listener.onResponse(AuthenticationResult.success(userAndPassword.user()));
 }
```
to
```
 if (userAndPassword == null || userAndPassword.passwordHash() == null) {
     listener.onResponse(AuthenticationResult.notHandled());
 } else if (userAndPassword.verify(password)) {
     listener.onResponse(AuthenticationResult.success(userAndPassword.user()));
 }
```
and the `verify()` method would then get the appropriate Hasher, based on the prefix of the stored hash. 
I'm just thinking this can't be used in the same way in `verifyPassword()` of `FileUserPasswdStore` where you have `Map<String, char[]>` instead of `UserAndPassword` entities, and in general I don't see the potential issues with the current approach ( apart from the extra Hasher object in each Store ). 

> I am leaning towards having a set of common alg+cost pairs and white list only these.

I am actually now thinking that we can use a PBKDF2_XXXX format with a sensible format (similar to bcrypt implicit cost factor ) that would allow users to set the cost to arbitrary values as they see fit in their implementation. There is no point in timing the hashing function in advance as this will highly depend on HW and i.e. 200 msec might be ok for a deployment where token service is in use and password hashes are verified every X minutes instead of in every request.

albertzaharovits commented:
> I disagree. Users wanting to run in a FIPS 140 JVM will have to set the CACHE_HASH_ALGO_SETTING to PBKDF2 since this is the only one that is approved.
For all the rest of the users PBKDF2 or bcrypt for cached hashes can be extremely slow and this is not the algorithms intended use. If something, this would bring forward the risk of users setting the cost to something extremely low, thus making the stored hashes more vulnerable to offline brute force attacks ( which is what PBKDF are meant to make difficult )

Understood, you are right. Probably only folks working with FIPS will touch the CACHE_HASH_ALGO_SETTING and they will have to set it to PBKDF2.

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:
```diff
@@ -9,134 +9,189 @@
 import java.util.Random;
 
-public enum Hasher {
+public interface Hasher {
```
If we go with a list of predefined hash algorithms with costs, then we can set this back to an enum and have singletons that avoid extra allocations

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:
```diff
@@ -9,134 +9,189 @@
-            return BCrypt.hashpw(text, salt).toCharArray();
+        public char[] hash(SecureString data) {
+            try {
+                StringBuilder result = new StringBuilder();
```
rather than use a string builder, can we use a CharBuffer? This avoids the creation of the string of the hash that goes into the string table and instead we just keep it in a char[]

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:
```diff
@@ -182,108 +249,29 @@ public boolean verify(SecureString text, char[] hash) {
+    final class SaltProvider {
 
-    static final class SaltProvider {
+        private SaltProvider() {
```
👍

jaymode commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/XPackSettingsTests.java:
```diff
@@ -10,6 +10,7 @@
 
 import javax.crypto.Cipher;
 
+import static org.hamcrest.Matchers.equalTo;
```
not needed?

tvernum commented:
I realised we've moved on from the previous implementation, but to circle back to this:

> The threat model doesn't usually include offline brute force attacks against the hashed passwords in the cache, thus I assumed the default cost would be sufficient for this.

I was thinking the reverse direction. The default cost is too high for the cache. If the thread model determines that a cost of 10,000 is appropriate for long term storage, then it is almost certainly too slow for use in the cache.

jkakavas commented:
> I was thinking the reverse direction. The default cost is too high for the cache. If the thread model determines that a cost of 10,000 is appropriate for long term storage, then it is almost certainly too slow for use in the cache.

Fair point, thanks. It will be configurable now so we can work around this.

jkakavas commented:
@albertzaharovits , @jaymode I addressed your feedback , this is ready for another round of review. 

Since I went with Albert's suggestion on verifying password hashes based on the stored hash and not on the configuration option, I removed the need of passing around algorithm setting in integration tests and select a random one for creating users / changing passwords.

jkakavas commented:
CI Run failed for unrelated reasons ( similar to #31116 ), so please test this @elasticmachine

albertzaharovits commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/ChangePasswordRequestBuilder.java:
```diff
@@ -41,22 +41,23 @@ public ChangePasswordRequestBuilder username(String username) {
 
-    public static char[] validateAndHashPassword(SecureString password) {
+    public static char[] validateAndHashPassword(SecureString password, String hashingAlgorithm) {
+        final Hasher hasher = Hasher.resolve(hashingAlgorithm, Hasher.BCRYPT);
```
I suggest we use the `Hasher` enum in place of the `String` value of the setting. Meaning, resolve to enum asap. In this case, resolve enum in `RestChangePasswordAction`'s ctr .
Also, I don't think resolve requires a default value, as there is a default for the setting value.

albertzaharovits replied:
I wonder if there should be a `public static <T extends Enum<T>> Setting<T> Setting#enumSetting(String key, Class<T> class, Function<Settings, T> defaultValue)`

albertzaharovits commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:
```diff
@@ -136,6 +144,182 @@ public boolean verify(SecureString text, char[] hash) {
+                return false;
+            }
+            return BCrypt.checkpw(text, hashStr);
+        }
```
Maybe factor these `#verify` as private static functions, similar to `verifyPbkdf2Hash`.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java:
```diff
@@ -73,9 +72,14 @@ public ReservedRealm(Environment env, Settings settings, NativeUsersStore native
-        final char[] hash = BOOTSTRAP_ELASTIC_PASSWORD.get(settings).length() == 0 ? EMPTY_PASSWORD_HASH :
-                Hasher.BCRYPT.hash(BOOTSTRAP_ELASTIC_PASSWORD.get(settings));
-        bootstrapUserInfo = new ReservedUserInfo(hash, true, hash == EMPTY_PASSWORD_HASH);
+
```
nit: extra line

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/ReservedRealm.java:
```diff
@@ -91,15 +95,15 @@ protected void doAuthenticate(UsernamePasswordToken token, ActionListener<Authen
                         if (userInfo.hasEmptyPassword) {
                             result = AuthenticationResult.terminate("failed to authenticate user [" + token.principal() + "]", null);
-                        } else if (Hasher.BCRYPT.verify(token.credentials(), userInfo.passwordHash)) {
+                        } else if (Hasher.verifyHash(token.credentials(), userInfo.passwordHash)) {
```
I think `UserInfo` can also have a `#verifyPassword` just as `UserAndPassword`.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/UserAndPassword.java:
```diff
@@ -34,6 +36,16 @@ public User user() {
+        } catch (IllegalArgumentException e) {
+            // The hash was of invalid format, we can't validate the password
+            return false;
+        }
```
I think `hasher`can be resolved in the constructor and throwing `IllegalArgumentException` theere is fine.

albertzaharovits commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:
```diff
@@ -265,6 +475,74 @@ public static Hasher resolve(String name, Hasher defaultHasher) {
+            result.put(Base64.getEncoder().encodeToString(secretKeyFactory.generateSecret(keySpec).getEncoded()));
+            return result.array();
+        } catch (InvalidKeySpecException | NoSuchAlgorithmException e) {
+            throw new IllegalStateException("Can't use PKDF2 for password hashing", e);
```
nit: PBKDF2
I think you can throw `ElasticSearchException`. I think this is the goto unchecked exception unless smth else fits better.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/user/RestChangePasswordAction.java:
```diff
@@ -61,7 +62,8 @@ public RestChannelConsumer innerPrepareRequest(RestRequest request, NodeClient c
                 new SecurityClient(client)
-                        .prepareChangePassword(username, request.requiredContent(), request.getXContentType())
+                    .prepareChangePassword(username, request.requiredContent(), request.getXContentType(), XPackSettings
+                        .PASSWORD_HASHING_ALGORITHM.get(settings))
```
I suggest we resolve this once in the ctr, and pass down the enum.

jkakavas replied:
Agreed, will do the same for `RestPutUserAction`

jkakavas commented:
Thanks @albertzaharovits. Appreciated the feedback, I addressed all your suggestions

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/user/RestChangePasswordAction.java:
```diff
@@ -57,11 +59,12 @@ public RestChannelConsumer innerPrepareRequest(RestRequest request, NodeClient c
         }
-
+        final Hasher passwordHasher = Hasher.resolve(XPackSettings
+            .PASSWORD_HASHING_ALGORITHM.get(settings));
```
whim: make it instance variable

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:
```diff
@@ -261,29 +421,108 @@ public static Hasher resolve(String name, Hasher defaultHasher) {
+            result.put(PBKDF2_PREFIX);
+            result.put(String.valueOf(cost));
+            result.put("$");
+            char[] salt = SaltProvider.salt(32);  // 32 characters for 64 bytes
```
My recollection was that this salt provider was enough for our `ssha` implementation that is only used for the cache. I have concerns about it as a provider for stored hashes. The BCrypt implementation provides its own salt so we do not use this implementation there. I'd think it would be best to allocate a 64 byte array and fill the bytes with a securerandom. Then we can base64 encode those bytes as the salt. What do you think?

jkakavas replied:
I incidentally had this discussion with @talevy yesterday when I was telling him how the SaltProvider uses SecureRandom now in the process of this discussion we figured out that this change didn't make it to the PR somehow. 
I think the entropy of the generated salt will be the same in your approach and in the current one as both will generate a 32 character string from an alphabet of 64 characters. That said, in the spirit of removing code, I have no objections following your suggestion.

jaymode replied:
Are you sure the entropy is the same? With the salt provider, the number of possible values for each position is 64. Meanwhile for actual bytes the value could be between 0-255 (256 possible values). I think the extra chars from base64 encoding is ok.

```java
SaltProvider.salt(32);  // 32 characters for 64 bytes
```

Given the value of the chars, each char fits in a single byte so we aren't really getting 64 bytes. (I am pretty sure about this, I didn't test though).

jkakavas replied:
> Given the value of the chars, each char fits in a single byte so we aren't really getting 64 bytes. (I am pretty sure about this, I didn't test though).

You're absolutely right, this tripped me. All chars used actually need 1 byte. 

> Are you sure the entropy is the same? 

Yes, but maybe I stated it incorrectly - or misused `entropy`.  My argument was that if by using either of the two approaches we end up with a char array of length 64 as a salt, so 512 bits, to exhaust that space one would need 2^512 attempts.

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:
```diff
@@ -261,29 +421,108 @@ public static Hasher resolve(String name, Hasher defaultHasher) {
+    private static boolean verifyPbkdf2Hash(SecureString data, char[] hash) {
+        try {
+            String hashStr = new String(hash);
+            if (!hashStr.startsWith(PBKDF2_PREFIX)) {
```
coding style wise the team is generally using `== false` instead of `!`

jkakavas replied:
I am aware, old habits die hard though.

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:
```diff
@@ -261,29 +421,108 @@ public static Hasher resolve(String name, Hasher defaultHasher) {
+            if (!hashStr.startsWith(PBKDF2_PREFIX)) {
+                return false;
+            }
+            hashStr = hashStr.replace(PBKDF2_PREFIX, "");
```
I think it would be better to use substring instead of replace. This will avoid the full scan of the string

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:
```diff
@@ -261,29 +421,108 @@ public static Hasher resolve(String name, Hasher defaultHasher) {
+        return BCrypt.checkpw(text, hashStr);
+    }
+
+    public static List<String> getAvailableAlgoStoredHash() {
```
let's add a javadoc

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:
```diff
@@ -261,29 +421,108 @@ public static Hasher resolve(String name, Hasher defaultHasher) {
+    }
+
+    private static boolean verifyBcryptHash(SecureString text, char[] hash) {
+        String hashStr = new String(hash);
```
Not new here but this is brought to light by this PR, we should avoid creating a new string from the hash. Instead we can use `CharArrays#charsBeginsWith`

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:
```diff
@@ -261,29 +421,108 @@ public static Hasher resolve(String name, Hasher defaultHasher) {
+
+    private static boolean verifyPbkdf2Hash(SecureString data, char[] hash) {
+        try {
+            String hashStr = new String(hash);
```
Let's see if we can avoid copying the hash into a string. If we do this will invalidate some of my earlier comments

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java:
```diff
@@ -75,20 +76,24 @@
  */
 public class NativeUsersStore extends AbstractComponent {
 
+
```
not needed new line

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/UserAndPassword.java:
```diff
@@ -34,6 +38,10 @@ public User user() {
     }
 
+    boolean verifyPassword(SecureString data) {
+            return hasher.verify(data, this.passwordHash);
```
indentation is off

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/file/FileUserPasswdStore.java:
```diff
@@ -84,14 +82,14 @@ public AuthenticationResult verifyPassword(String username, SecureString passwor
 
     public boolean userExists(String username) {
-        return users.containsKey(username);
+        return users != null && users.containsKey(username);
```
why the null check? I specifically changed this a little while ago

jkakavas replied:
I have no specific recollection, I assume this was leftover from an earlier approach/change that was discarded. Will revert.

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/user/RestChangePasswordAction.java:
```diff
@@ -32,6 +34,8 @@
 public class RestChangePasswordAction extends SecurityBaseRestHandler implements RestRequestFilter {
 
     private final SecurityContext securityContext;
+    private final Hasher passwordHasher = Hasher.resolve(XPackSettings
```
is the split of lines needed?

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/user/RestPutUserAction.java:
```diff
@@ -34,6 +36,9 @@
  */
 public class RestPutUserAction extends SecurityBaseRestHandler implements RestRequestFilter {
 
+    private final Hasher passwordHasher = Hasher.resolve(XPackSettings
```
is the split of lines needed?

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/integration/ClearRealmsCacheTests.java:
```diff
@@ -42,10 +41,11 @@
 
     private static String[] usernames;
 
+
```
not needed new lines

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/integration/ClusterPrivilegeTests.java:
```diff
@@ -74,7 +71,13 @@ protected String configRoles() {
     @Override
     protected String configUsers() {
-        return super.configUsers() + USERS;
+        final String usersPasswdHashed = new String(Hasher.resolve(randomFrom("pbkdf2", "pbkdf2_1000", "bcrypt",
```
maybe avoid the line split by moving out the randomFrom into its own line?

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/integration/DocumentLevelSecurityRandomTests.java:
```diff
@@ -27,17 +26,19 @@
 
     protected static final SecureString USERS_PASSWD = new SecureString("change_me".toCharArray());
-    protected static final String USERS_PASSWD_HASHED = new String(Hasher.BCRYPT.hash(new SecureString("change_me".toCharArray())));
+
```
can drop the new line

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/integration/DocumentLevelSecurityTests.java:
```diff
@@ -81,7 +80,7 @@
 
     protected static final SecureString USERS_PASSWD = new SecureString("change_me".toCharArray());
-    protected static final String USERS_PASSWD_HASHED = new String(Hasher.BCRYPT.hash(USERS_PASSWD));
+
```
can drop the new line

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/integration/FieldLevelSecurityTests.java:
```diff
@@ -67,7 +66,8 @@
 
     protected static final SecureString USERS_PASSWD = new SecureString("change_me".toCharArray());
-    protected static final String USERS_PASSWD_HASHED = new String(Hasher.BCRYPT.hash(new SecureString("change_me".toCharArray())));
+
```
can drop the new lines

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/integration/IndexPrivilegeTests.java:
```diff
@@ -129,7 +114,24 @@ protected String configRoles() {
     @Override
     protected String configUsers() {
-        return super.configUsers() + USERS;
+        final String usersPasswdHashed = new String(Hasher.resolve(randomFrom("pbkdf2", "pbkdf2_1000", "bcrypt",
```
let's see if we can avoid the wrap by moving the randomFrom to its own line

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/integration/KibanaUserRoleIntegTests.java:
```diff
@@ -55,16 +54,22 @@ public String configRoles() {
                 "my_kibana_user:kibana_user\n" +
                 "kibana_user:kibana_user";
     }
+    @Override
```
looks like this change can be removed?

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/integration/KibanaUserRoleIntegTests.java:
```diff
@@ -55,16 +54,22 @@ public String configRoles() {
 
     @Override
     public String configUsersRoles() {
+
```
unnecessary new line

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/integration/PermissionPrecedenceTests.java:
```diff
@@ -51,9 +49,11 @@ protected String configRoles() {
-        return "admin:" + USERS_PASSWD_HASHED + "\n" +
-                "client:" + USERS_PASSWD_HASHED + "\n" +
-                "user:" + USERS_PASSWD_HASHED + "\n";
+        final String usersPasswdHashed = new String(getFastStoredHashAlgoForTests().
```
can we avoid the line split?

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/integration/SecurityClearScrollTests.java:
```diff
@@ -33,15 +32,16 @@
 
     @Override
     protected String configUsers() {
+        final String usersPasswdHashed = new String(getFastStoredHashAlgoForTests().
```
can we avoid the line split?

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/test/SecurityIntegTestCase.java:
```diff
@@ -77,7 +78,7 @@
 public abstract class SecurityIntegTestCase extends ESIntegTestCase {
 
-    private static SecuritySettingsSource SECURITY_DEFAULT_SETTINGS;
+    protected static SecuritySettingsSource SECURITY_DEFAULT_SETTINGS;
```
looks like this doesn't need to be `protected`

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/test/SecuritySettingsSource.java:
```diff
@@ -133,7 +135,7 @@ public Settings nodeSettings(int nodeOrdinal) {
                 .put("xpack.security.authc.realms.file.order", 0)
                 .put("xpack.security.authc.realms.index.type", NativeRealmSettings.TYPE)
-                .put("xpack.security.authc.realms.index.order", "1");
+            .put("xpack.security.authc.realms.index.order", "1");
```
formatting is off. Let's line this up with everything else

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/user/TransportChangePasswordActionTests.java:
```diff
@@ -54,10 +54,10 @@ public void testAnonymousUser() {
         ChangePasswordRequest request = new ChangePasswordRequest();
         request.username(anonymousUser.principal());
-        request.passwordHash(Hasher.BCRYPT.hash(SecuritySettingsSourceField.TEST_PASSWORD_SECURE_STRING));
+        request.passwordHash(Hasher.resolve(randomFrom("pbkdf2", "pbkdf2_1000", "bcrypt", "bcrypt9")).
```
can we avoid the line breaks introduced here?

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/user/TransportChangePasswordActionTests.java:
```diff
@@ -85,10 +85,10 @@ public void testInternalUsers() {
         ChangePasswordRequest request = new ChangePasswordRequest();
         request.username(randomFrom(SystemUser.INSTANCE.principal(), XPackUser.INSTANCE.principal()));
-        request.passwordHash(Hasher.BCRYPT.hash(SecuritySettingsSourceField.TEST_PASSWORD_SECURE_STRING));
+        request.passwordHash(Hasher.resolve(randomFrom("pbkdf2", "pbkdf2_1000", "bcrypt", "bcrypt9")).
```
can we avoid the line break introduced here

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/user/TransportChangePasswordActionTests.java:
```diff
@@ -153,7 +154,8 @@ public void testException() {
         ChangePasswordRequest request = new ChangePasswordRequest();
         request.username(user.principal());
-        request.passwordHash(Hasher.BCRYPT.hash(SecuritySettingsSourceField.TEST_PASSWORD_SECURE_STRING));
+        request.passwordHash(Hasher.resolve(randomFrom("pbkdf2", "pbkdf2_1000", "bcrypt", "bcrypt9"))
```
can we avoid the line break introduced here

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/user/TransportPutUserActionTests.java:
```diff
@@ -167,7 +167,8 @@ public void testValidUser() {
         request.username(user.principal());
         if (isCreate) {
-            request.passwordHash(Hasher.BCRYPT.hash(SecuritySettingsSourceField.TEST_PASSWORD_SECURE_STRING));
+            request.passwordHash(Hasher.resolve(randomFrom("pbkdf2", "pbkdf2_1000", "bcrypt", "bcrypt9")).
```
would like to avoid the line break

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/RealmSettingsTests.java:
```diff
@@ -261,7 +260,7 @@ public void testSettingsWithMultipleRealmsValidatesSuccessfully() throws Excepti
             builder.put("cache.ttl", randomPositiveTimeValue())
                     .put("cache.max_users", randomIntBetween(1_000, 1_000_000))
-                    .put("cache.hash_algo", randomFrom(HASH_ALGOS));
+                .put("cache.hash_algo", randomFrom(CACHE_HASHING_ALGOS));
```
indentation is off here. Either way works for me with regards to changing one line vs all of them

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/file/FileUserPasswdStoreTests.java:
```diff
@@ -55,6 +55,8 @@ public void init() {
         settings = Settings.builder()
                 .put("resource.reload.interval.high", "2s")
                 .put("path.home", createTempDir())
+            .put("xpack.security.authc.password_hashing.algorithm", randomFrom("bcrypt", "bcrypt11", "pbkdf2", "pbkdf2_1000",
```
move the whole randomFrom to its own line to make it easier to read

jkakavas commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/ChangePasswordRequestBuilder.java:
```diff
@@ -41,22 +41,22 @@ public ChangePasswordRequestBuilder username(String username) {
      * Sets the password. Note: the char[] passed to this method will be cleared.
      */
-    public ChangePasswordRequestBuilder password(char[] password) {
+    public ChangePasswordRequestBuilder password(char[] password, Hasher hasher) {
```
is or _is not_ ? I Can definitely reintroduce the `ChangePasswordRequestBuilder password(char[] password)` signature and call `ChangePasswordRequestBuilder password(char[] password, Hasher hasher)` with a default Hasher instance (`BCrypt`)

jaymode replied:
Sorry for the incomplete thought. I think breaking is the right way to go; the issue with a non-breaking change is the default method will be trappy if they configure elasticsearch to use PBKDF2 but we hash with BCrypt on the client side by default. I wonder if we should add validation that the hash is the correct type in TransportChangePasswordAction?

jkakavas commented:
@jaymode this is ready for a new round. Wasn't sure about adding the check to `TransportChangePasswordAction` as discussed above. No objections adding this, I just wonder if it is necessary if we go for the breaking change.

jaymode commented on server/src/main/java/org/elasticsearch/common/settings/Setting.java:
```diff
@@ -1017,6 +1017,14 @@ public void apply(T value, Settings current, Settings previous) {
+     * @param key          the settings key for this setting.
+     * @param defaultValue the default String value.
+     * @param properties   properties for this setting like scope, filtering...
+     * @return
```
remove empty return in javadocs

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:
```diff
@@ -472,35 +493,46 @@ public static boolean verifyHash(SecureString data, char[] hash) {
     private static boolean verifyPbkdf2Hash(SecureString data, char[] hash) {
+        // Base64 string length : (4*(n/3)) rounded up to the next multiple of 4 because of padding,  i.e. 44 for 32 bytes
+        final int tokenLength = 44;
+        char[] hashChars = new char[tokenLength];
```
lets set these to null. We never use these arrays. We can have null checks in the finally block.

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:
```diff
@@ -472,35 +493,46 @@ public static boolean verifyHash(SecureString data, char[] hash) {
+                cost, PBKDF2_KEY_LENGTH);
+            char[] computedPwdHash = CharArrays.utf8BytesToChars(Base64.getEncoder()
+                .encode(secretKeyFactory.generateSecret(keySpec).getEncoded()));
+            boolean result = CharArrays.constantTimeEquals(computedPwdHash, hashChars);
```
make it final

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:
```diff
@@ -511,27 +543,14 @@ private static boolean verifyBcryptHash(SecureString text, char[] hash) {
+     * Generates an array of {@code length} random bytes using {@link java.security.SecureRandom}
+     */
+    private static byte[] generateSalt(int length) {
+        Random random = new SecureRandom();
```
can we store the value? SecureRandom objects can be reused and we don't need to keep creating them

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/PasswordHashingAlgorithmBootstrapCheck.java:
```diff
@@ -0,0 +1,47 @@
+public class PasswordHashingAlgorithmBootstrapCheck implements BootstrapCheck {
+    @Override
+    public BootstrapCheckResult check(BootstrapContext context) {
+        final String selectedAlgorithm = XPackSettings.PASSWORD_HASHING_ALGORITHM.get(context.settings);
```
If I told you to move this here, I am sorry. Setting validation should happen on the setting and not be part of a bootstrap check. The availability of the PBKDF2 algorithm is fine as a bootstrap check

jkakavas replied:
I misread the original comment, I'll revert the setting validation and only check the algo availability here.

jaymode commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:
```diff
@@ -285,7 +285,8 @@ public Security(Settings settings, final Path configPath) {
                     new TokenSSLBootstrapCheck(),
                     new PkiRealmBootstrapCheck(settings, getSslService()),
-                    new TLSLicenseBootstrapCheck()));
+                new TLSLicenseBootstrapCheck(),
```
indentation needs to be cleaned up

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:
```diff
@@ -506,17 +512,21 @@ private static boolean verifyPbkdf2Hash(SecureString data, char[] hash) {
             SecretKeyFactory secretKeyFactory = SecretKeyFactory.getInstance("PBKDF2withHMACSHA512");
             PBEKeySpec keySpec = new PBEKeySpec(data.getChars(), Base64.getDecoder().decode(CharArrays.toUtf8Bytes(saltChars)),
-                cost, PBKDF2_KEY_LENGTH);
+                cost, 256);
```
I like the constant more since it provided context to this "magic number"

jkakavas replied:
I eventually removed it since I didn't use it in calculating the 44 "magic number" in code and it seemed redundant. I can make it more obvious in the comment that `n` in `(4*(n/3))` is 32 because its the key size (256) in bytes

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java:
```diff
@@ -506,17 +512,21 @@ private static boolean verifyPbkdf2Hash(SecureString data, char[] hash) {
                 .encode(secretKeyFactory.generateSecret(keySpec).getEncoded()));
-            boolean result = CharArrays.constantTimeEquals(computedPwdHash, hashChars);
+            final boolean result = CharArrays.constantTimeEquals(computedPwdHash, hashChars);
             Arrays.fill(computedPwdHash, '\u0000');
```
not a big deal, but maybe we should pull the `computedPwdHash` array out of the try and set it to null initially. Then we can fill it in the finally with the others?

jkakavas replied:
I was just trying to zerofill as soon after the use as possible. Do you see any advantages other than consistency?

jaymode replied:
An exception is thrown during the conversion of utf8 bytes to chars would cause this array to hang around in memory. Like I said, if you prefer how you have it, I am fine with it :)

jkakavas replied:
Gotcha. Nope, I was just trying to see what you're getting at thanks. I'll change it

jaymode commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/PasswordHashingAlgorithmBootstrapCheckTests.java:
```diff
@@ -17,7 +17,7 @@
         Settings settings = Settings.EMPTY;
         assertFalse(new PasswordHashingAlgorithmBootstrapCheck().check(new BootstrapContext(settings, null)).isFailure());
-
+        // The following two will always pass because for now we only test in environments where PBKDF2WithHMACSHA512 is supported
```
can we add an `assume` statement that validates PBKDF2 is available?
